crates/oxc_linter/src/ast_util.rs

```diff
@@ -975,3 +975,149 @@ pub fn is_node_within_call_argument<'a>(
+/// * Alphanumeric characters, `_`, `$` - Identifiers
+/// * `"`, `'`, `` ` `` - String/template literals
+/// * `++`, `--` - Postfix operators
+///
+/// # Example
+///
+/// ```ignore
+/// ctx.diagnostic_with_fix(diagnostic, |fixer| {
+///     let needs_semi = could_be_asi_hazard(node, ctx);
+///     let prefix = if needs_semi { ";" } else { "" };
+///     fixer.replace(call_expr.span, format!("{prefix}[...foo]"))
+/// });
+/// ```
+pub fn could_be_asi_hazard(node: &AstNode, ctx: &LintContext) -> bool {
+    let node_span = node.span();
+
+    // Walk up ancestors to find an ExpressionStatement.
+    // We only care about ExpressionStatements because that's the only context
+    // where ASI hazards can occur - other statements like `return`, `const`, etc.
+    // have their own syntax that prevents ambiguity.
+    let Some(expr_stmt_span) = ctx.nodes().ancestors(node.id()).find_map(|ancestor| {
+        if let AstKind::ExpressionStatement(expr_stmt) = ancestor.kind() {
+            Some(expr_stmt.span)
+        } else {
+            None
+        }
+    }) else {
+        return false;
+    };
+
+    // Check if our node starts at the same position as the ExpressionStatement.
+    // If not, our replacement won't be at the start of the statement, so no ASI hazard.
+    // For example: `foo.bar(Array.from(x))` - the Array.from is not at statement start.
+    if node_span.start != expr_stmt_span.start {
+        return false;
+    }
+
+    // Now check if the character before the ExpressionStatement could cause ASI issues
+    if expr_stmt_span.start == 0 {
+        return false;
+    }
+
+    let source_text = ctx.source_text();
+
+    // Find the last meaningful character before the span, skipping whitespace and comments.
+    let last_char = find_last_meaningful_char(source_text, expr_stmt_span.start, ctx);
+
+    let Some(last_char) = last_char else {
+        return false;
+    };
+
+    // These characters, when followed by `[`, `(`, `/`, `+`, `-`, or `` ` ``,
+    // could cause the new code to be parsed as a continuation of the previous expression
+    matches!(last_char, ')' | ']' | '}' | '"' | '\'' | '`' | '+' | '-')
+        || last_char.is_alphanumeric()
+        || last_char == '_'
+        || last_char == '$'
+}
+
+/// Find the last meaningful (non-whitespace, non-comment) character before the given position.
+/// Uses semantic's comment tracking to skip entire comment spans efficiently.
+fn find_last_meaningful_char(source_text: &str, end_pos: u32, ctx: &LintContext) -> Option<char> {
+    let bytes = source_text.as_bytes();
+    let mut i = end_pos;
+
+    while i > 0 {
+        i -= 1;
+
+        let byte = bytes[i as usize];
+
+        // Skip ASCII whitespace (single byte)
+        if byte.is_ascii_whitespace() {
+            continue;
+        }
+
+        if let Some(comment) = ctx.semantic().get_comment_at(i) {
+            i = comment.span.start;
+            continue;
+        }
```

Comment:
Oh actually I missed a much bigger perf problem. `.get_comment_at()` is really expensive, and it's being called over and over for every byte.

What would be better is:

1. Before the loop:
  * Get the index of last comment that starts before `end_pos`.
  * If `end_pos` is inside that comment, jump forwards to start of that comment, and get the previous comment.
  * Store `comment.span` (or 0 span if no comment).
2. Main loop:
  * Loop through bytes while `i >= comment_span.end`, checking them.
3. After main loop:
  * Skip forwards to `comment_span.start`.
  * Fetch previous comment.
4. Go back to step 2.

i.e. binary search comments only once, and thereafter step through comments 1-by-1 in reverse order.